bdc_collection_builder/collections/sentinel/tasks.py

```diff
@@ -277,6 +280,74 @@ def upload(self, scene):
             logging.warning('Uploading {} to BUCKET {} - {}'.format(entry['file'], Config.AWS_BUCKET_NAME, file_without_prefix))
             upload_file(entry['file'], Config.AWS_BUCKET_NAME, file_without_prefix)
 
+    def harmonize(self, scene):
+        """Apply Harmonization on Sentinel-2 collection.
+
+        Args:
+            scene - Serialized Activity
+        """
+        # Set Collection to the Sentinel NBAR (Nadir BRDF Adjusted Reflectance)
+        scene['collection_id'] = 'S2NBAR'
+        scene['activity_type'] = 'harmonizeS2'
+
+        # Create/Update activity
+        activity_history = self.create_execution(scene)
+
+        logging.debug('Starting Harmonization Sentinel...')
+
+        activity_history.activity.activity_type = 'harmonizeS2'
+        activity_history.start = datetime.utcnow()
+        activity_history.save()
+
+        try:
+            SAFEL2A = scene['args']['file']
+
+            # Define new filenames for products
+            parts = os.path.basename(SAFEL2A).split('_')
+
+            # Get year month from .SAFE folder
+            ymd_part = parts[2]
+            y = ymd_part[:4]
+            m = ymd_part[4:6]
+            d = ymd_part[6:8]
+            yyyymm = '{}-{}'.format(y, m)
+            mgrs = parts[5]
+
+            dir_published_L2 = str(Path(Config.DATA_DIR) / 'Repository/Archive/{}/{}/{}/'.format('S2SR_SEN28', yyyymm, os.path.basename(SAFEL2A)))
+            L1_dir = str(Path(Config.DATA_DIR) / 'Repository/Archive/{}/{}/'.format('S2_MSI', yyyymm))
+            zip_pattern = '.*_MSIL1C_{}{}{}.*_{}_.*.zip$'.format(y,m,d,mgrs)
+            zip_file_name = [os.path.join(L1_dir,d) for d in os.listdir(L1_dir) if re.match('{}'.format(zip_pattern), d)][0]
+
+            # Check if file is valid
+            valid = is_valid(zip_file_name)
+
+            if not valid:
+                raise IOError('Invalid zip file "{}"'.format(zip_file_name))
```

Comment:
This section looks like the `download` step where we must check if zip is valid before extract.
Could you create a common function to these code?

Reply:
done (extract_and_get_internal_name)